src/coreclr/vm/prestub.cpp

```diff
@@ -1059,7 +1045,437 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, JitListLockEn
+    }
+
+    // Using the kind type, perform the following:
+    //  1) Validate the basic type information from the signature.
+    //  2) Resolve the name to the appropriate member.
+    switch (context.Kind)
+    {
+    case UnsafeAccessorKind::Constructor:
+        // A return type is required for a constructor, otherwise
+        // we don't know the type to construct.
+        // The name is defined by the runtime and should be empty.
+        if (context.DeclarationSig.IsReturnTypeVoid() || retType.IsNull() || !name.IsEmpty())
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSOR);
+        }
+
+        context.TargetType = retType;
+        if (!TrySetTargetMethodCtor(context))
+            MemberLoader::ThrowMissingMethodException(firstArgType.GetMethodTable(), ".ctor");
+        break;
+
+    case UnsafeAccessorKind::Method:
+    case UnsafeAccessorKind::StaticMethod:
+        // Method access requires a target type.
+        if (firstArgType.IsNull())
+            ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSOR);
+
+        context.TargetType = firstArgType;
+        context.IsTargetStatic = kind == UnsafeAccessorKind::StaticMethod;
+        if (!TrySetTargetMethod(context, name.GetUTF8()))
+            MemberLoader::ThrowMissingMethodException(firstArgType.GetMethodTable(), name.GetUTF8());
+        break;
+
+    case UnsafeAccessorKind::Field:
+    case UnsafeAccessorKind::StaticField:
+        // Field access requires a single argument for target type and a return type.
+        if (argCount != 1 || firstArgType.IsNull() || context.DeclarationSig.IsReturnTypeVoid() || retType.IsNull())
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSOR);
+        }
+
+        // The return type must be byref.
+        // If the non-static field access is for a
+        // value type, the instance must be byref.
+        if (!retType.IsByRef()
+            || (kind == UnsafeAccessorKind::Field
+                && firstArgType.IsValueType()
+                && !firstArgType.IsByRef()))
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSOR);
+        }
+
+        context.TargetType = firstArgType;
+        context.IsTargetStatic = kind == UnsafeAccessorKind::StaticField;
+        if (!TrySetTargetField(context, name.GetUTF8(), retType.GetTypeParam()))
+            MemberLoader::ThrowMissingFieldException(firstArgType.GetMethodTable(), name.GetUTF8());
+        break;
+
+    default:
+        ThrowHR(COR_E_BADIMAGEFORMAT, BFA_INVALID_UNSAFEACCESSOR);
+    }
+
+    // Due to how some types degrade, unsafe access on these
+    // target types is blocked, even if the lookup succeeds.
+    // For example, pointers can become lookups on typeof(uint)'s MethodTable.
+    // Regardless of lookup success, we are going to block
+    // unsafe access for some types.
+    if (context.TargetType.IsArray()
+        || context.TargetType.IsPointer())
```

Comment:
```suggestion
        || context.TargetType.IsTypeDesc())
```
This covers e.g. Constructor + function pointer that is invalid. 

I think it may be better to reject the invalid target types early before the method lookup.

Reply:
> What does the TypeDesc help with here? Is this a "built-in" vs user defined mechanism?

Found it in the BOTR.

Reply:
According to the doc, arrays are represented by `TypeDesc`s also, so perhaps that is all that is needed. I will check it out.

Reply:
> According to the doc, arrays are represented by TypeDescs also, so perhaps that is all that is needed

They used to be, but they are not anymore. Could you please fix the place in the docs where you have seen it?

I think it would ok to allow UnsafeAccessorAttribute with arrays. They would work just fine.

It is all the typedesc types (pointers, function pointers, generic variables) that are problematic because of they are not types with methods.

Reply:
> They used to be, but they are not anymore. Could you please fix the place in the docs where you have seen it?

Done.

> It is all the typedesc types (pointers, function pointers, generic variables) that are problematic because of they are not types with methods.

This makes sense.

Reply:
> I think it may be better to reject the invalid target types early before the method lookup.

Using `IsTypeDesc()` is going to make this complicated so late. Will move this elsewhere.